scipy/_lib/_array_api_util.py

```diff
@@ -0,0 +1,65 @@
+"""Utility functions to use Python Array API compatible libraries.
+
+For the context about the Array API see:
+https://data-apis.org/array-api/latest/purpose_and_scope.html
+
+The SciPy use case of the Array API is described on the following page:
+https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy
+"""
+
+import numpy as np
+
+from scipy._lib import _pep440
+
+_NUMPY_INCLUDES_ARRAY_API = _pep440.parse(np.__version__) >= _pep440.Version(
+    "1.22.0"
+)
+
+
+def _get_namespace(*arrays):
+    """
+    Returns the module that implements Array API compatible functions or NumPy.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays to get the namespace from.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> x = np.arange(6)
+    >>> _get_namespace(x).__name__
+    'numpy'
+
+    >>> import numpy.array_api as xp
+    >>> x = xp.arange(6)
+    >>> _get_namespace(x).__name__
+    'numpy.array_api'
+    """
+    # `arrays` contains one or more arrays
+    # When there is no __array_namespace__,
+    # we will use assume the default (NumPy) namespace.
+    namespaces = {
+        x.__array_namespace__() if hasattr(x, "__array_namespace__") else np
+        for x in arrays
+    }
+
+    if len(namespaces) != 1:
```

Comment:
I agree here. Better to fail explicitly than trying out an arbitrary conversion from one backend to another. Otherwise we will have reports of weird behaviours and performances.

Reply:
One thing we can consider, trying to convert lists into either np or the only other backend which is present.

Reply:
That might lead to confusions, because the code will work and the user would think that they are receiving the optimum performance but that might not be the case as always. IMO, failing is the best option in case of conflicting namespaces.

Reply:
The issue is that it would introduce a backward incompatibility in the sense that some user expect the function to work when they pass a list instead of an array of some type.

We could also add a (deprecation) warning instead of failing in that case so user can migrate their code.

Reply:
> We could also add a (deprecation) warning instead of failing in that case so user can migrate their code.

Yes. Though, eventually we should fail in case of conflicting namespaces. For migrating to such failures, as you said, we can use `DeprecationWarning` for a better user experience.